hydra/utils.py

```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
     :return: if _target_ is a class name: the instantiated object
              if _target_ is a callable: the return value of the call
     """
-
-    if OmegaConf.is_none(config):
+    # Return None if config is None
+    if config is None or OmegaConf.is_none(config):
         return None
 
-    if isinstance(config, TargetConf) and config._target_ == "???":
+    # TargetConf edge case
+    if isinstance(config, TargetConf) and config.is_missing():
         # Specific check to give a good warning about failure to annotate _target_ as a string.
         raise InstantiationException(
             f"Missing value for {type(config).__name__}._target_. Check that it's properly annotated and overridden."
             f"\nA common problem is forgetting to annotate _target_ as a string : '_target_: str = ...'"
         )
 
-    if not (
-        isinstance(config, dict)
-        or OmegaConf.is_config(config)
-        or is_structured_config(config)
-    ):
-        raise HydraException(f"Unsupported config type : {type(config).__name__}")
+    # Structured Config always converted first to OmegaConf
+    if is_structured_config(config):
+        config = OmegaConf.structured(config, flags={"allow_objects": True})
 
+    # Make shallow copy if config is a standard dictionary
     if isinstance(config, dict):
-        configc = config.copy()
-        _convert_container_targets_to_strings(configc)
-        config = configc
-
-    kwargsc = kwargs.copy()
-    _convert_container_targets_to_strings(kwargsc)
-    kwargs = kwargsc
+        config = config.copy()
 
-    # make a copy to ensure we do not change the provided object
-    config_copy = OmegaConf.structured(config, flags={"allow_objects": True})
+    # If OmegaConf, copy, make it mutable and set parent for live interpolation
     if OmegaConf.is_config(config):
+        config_copy = OmegaConf.structured(config, flags={"allow_objects": True})
+        OmegaConf.set_readonly(config_copy, False)
+        OmegaConf.set_struct(config_copy, False)
         config_copy._set_parent(config._get_parent())
-    config = config_copy
-    assert OmegaConf.is_config(config)
-    OmegaConf.set_readonly(config, False)
-    OmegaConf.set_struct(config, False)
+        config_copy._set_flag("allow_objects", True)
+        config = config_copy
 
-    target = _get_target_type(config, kwargs)
-    try:
-        config._set_flag("allow_objects", True)
-        final_kwargs = _get_kwargs(config, **kwargs)
-        convert = _pop_convert_mode(final_kwargs)
-        if convert == ConvertMode.PARTIAL:
-            final_kwargs = OmegaConf.to_container(
-                final_kwargs, resolve=True, exclude_structured_configs=True
+    # Finalize config (convert targets to strings, merge with kwargs)
+    # TODO: kwargs are used as a config now. If it contains dataclass
+    # objects, they will be interpreted as config nodes in the
+    # instantiation result. We might want to change that behavior.
```

Comment:
This is the main friction point.

I'd argue that this would be confusing to treat dataclass instances in overrides as objects and not configs.

1. Instantiate should operate on config-like objects, why would the overrides already have some instantiated objects?
2. DataClass become sources of confusion. If config is `{"obj": {"x": 1, "y": 2}}` and kwargs is `obj=SomeDataClass(x=2)`, what do we expect? `{"obj": {"x": 2, "y": 2}}` (current behavior), or `{"obj": SomeDataClass(x=2)`

Also, (not necessarily a good argument), it would make the implementation more complex. 
The current implementation does
1. merge kwargs fully into config
2. forget about kwargs, only look at config

If we want to treat kwargs as objects, we can't do that because dataclasses instances merged into OmegaConf containers would be converted back to OmegaConf containers. This means that we should keep the kwargs alonside the config during the whole instantiation path.